pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1079,26 +1082,28 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
 			return volumeToMount.GenerateError("MapVolume type assertion error", fmt.Errorf("volume host does not implement KubeletVolumeHost interface"))
 		}
 		hu := kvh.GetHostUtil()
-		devicePath, err = hu.EvalHostSymlinks(devicePath)
+		resolvedDevicePath, err := hu.EvalHostSymlinks(devicePath)
 		if err != nil {
-			return volumeToMount.GenerateError("MapVolume.EvalHostSymlinks failed", err)
+			// For CSI plugin, device won't be available at devicePath until MapPodDevice is called,
+			// so, just warn and continue.
+			klog.Warningf("MapVolume.EvalHostSymlinks failed for %s: %v", devicePath, err)
+		} else {
+			devicePath = resolvedDevicePath
 		}
 
-		// Map device to global and pod device map path
-		volumeMapPath, volName := blockVolumeMapper.GetPodDeviceMapPath()
-		mapErr = blockVolumeMapper.MapDevice(devicePath, globalMapPath, volumeMapPath, volName, volumeToMount.Pod.UID)
+		// Execute driver specific map
+		mapErr = blockVolumeMapper.MapPodDevice()
 		if mapErr != nil {
 			// On failure, return error. Caller will log and retry.
-			return volumeToMount.GenerateError("MapVolume.MapDevice failed", mapErr)
+			return volumeToMount.GenerateError("MapVolume.MapPodDevice failed", mapErr)
 		}
 
-		// Take filedescriptor lock to keep a block device opened. Otherwise, there is a case
-		// that the block device is silently removed and attached another device with same name.
-		// Container runtime can't handler this problem. To avoid unexpected condition fd lock
-		// for the block device is required.
-		_, err = og.blkUtil.AttachFileDevice(devicePath)
-		if err != nil {
-			return volumeToMount.GenerateError("MapVolume.AttachFileDevice failed", err)
+		// Execute common map
+		volumeMapPath, volName := blockVolumeMapper.GetPodDeviceMapPath()
+		mapErr = ioutil.MapBlockVolume(og.blkUtil, devicePath, globalMapPath, volumeMapPath, volName, volumeToMount.Pod.UID)
+		if mapErr != nil {
+			// On failure, return error. Caller will log and retry.
```

Comment:
This PR  creates `/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/<volume name>` and `/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/<volume name>`, both in global dir.

The spirit of NodeStage and NodePublish is to have Stage using global dir and Publish to use pod dir, not the global one.

I think that all in-tree volume plugins should perhaps have `ioutil.MapBlockVolume` in their respective MapPodDevice() plugin code, and CSI would call NodePublish **instead** of ioutil.MapBlockVolume.

In the end we should have the symlink (or bind-mount or whatever CSI driver uses) in the pod directory established by the driver instead of kubelet itself.

Reply:
Sure, but first ioutil.MapBlockVolume should have the code for the "refcount" globalMapPath extracted from it. And MapPodDevice should not take globalMapPath as an argument. The globalMapPath manipulation part of ioutil.MapBlockVolume is common to all plugins, it should not be a plugin responsibility and has been the source of a lot of confusion.

Reply:
Yeah, it may be harder to split things correctly, but IMO it's worth the effort. Having parallel structure to pod directory in the global directory looks really odd.

Reply:
I understand that the point of the issue is that both staging path and publish path exist under the global dir, not moving the common code out from the plugin code. I will consider moving publish path to other place (It might be 1. under pod dir and symlinked, 2. under pod dir and not symlinked, 3. not under pod dir and symlinked.)

Reply:
@jsafrane 

Could you review the last three commits? 
They are for making publish path == pod device map path (This means that CSI driver publishes directly to pod device map path).

The second commit is still not implemented in the right way, however I think that we can improve this by applying the similar approach to https://github.com/kubernetes/kubernetes/pull/84747/commits/ef964f0405b929767a58d0ee56eb9e78335a1371. (We would be able to further separate interface to one for `SetUpDevice`/`TearDownDevice` and one for `MapPodDevice`/`UnmapPodDevice` . Then, we will able to assume that plugins that implement later interface don't require common mapping/unmapping for pod path.)

If the above concept looks good, I will go ahead and implement this way.

Reply:
I dived deeper into the code and I start understanding why it's so complicated. We do not do NodeStage equivalent in in-tree plugins, we just put symlink **for every pod** both into global dir (for refcounting) and into pod local dir (for pod).

In this PR, you Stage the dir into a global dir (that's special for CSI) and then Publish in the global dir (for refcounting) and create a symlink in the pod dir (as usual). We can do the other way around (use symlink for refcounting and NodePublish into pod dir), but the result would be largely the same, potentially complicating the code.

I am sorry, the original PR may be actually better in this, as it requires less refactoring. I will go through the PR, however @davidz627 @msau42 @jingxu97, your input would be valuable too.

Reply:
@gnufied raised an issue regarding recovery handling for CSI raw blocks: https://github.com/kubernetes/kubernetes/issues/82340. I think that would be the most important issue to solve. Is that addressed here?

Reply:
@msau42 
> Is that addressed here?

Even without the last three commit, this PR will solve the issue. We are focusing on to solve the issue by this PR.

Reply:
To pull #74026, or the original PR, I've pushed the version without the last three commits in this PR on top of the latest prerequisite PRs. I hope it works and looks good.